In [1]:
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

import numpy as np
import scipy as sp
import math
from scipy.integrate import odeint
from scipy.stats import poisson
from scipy.stats import norm
from scipy.stats import lognorm

import os
import sys
import io
from contextlib import redirect_stdout

from repairanalysis import medrasrepair
from repairanalysis import sddparser
from itertools import groupby

path = os.getcwd()
sddpath = path

In [2]:
ret_dat = medrasrepair.repairSimulation(sddpath, 'postRepairDNA')
#print(type(ret_dat))
#print(ret_dat)

From folder:	 /content/ 

Data for: n5MeV_inner_alpha_3_SDDOutput.txt
SDR Version: SDRv2.0
Author: Zhenlun Dai
Associated SDD File: n5MeV_inner_alpha_3_SDDOutput.txt
Intact Chromosome Sizes: 46, 263.2, 258.1, 212.4, 203.4, 192.2, 181.8, 168.8, 155.5, 149.4, 144.0, 142.9, 140.8, 121.3, 113.0, 106.9, 94.7, 83.5, 81.0, 68.0, 66.6, 50.0, 52.6, 263.2, 258.1, 212.4, 203.4, 192.2, 181.8, 168.8, 155.5, 149.4, 144.0, 142.9, 140.8, 121.3, 113.0, 106.9, 94.7, 83.5, 81.0, 68.0, 66.6, 50.0, 52.6, 61.6, 165.2, 
***end of master header***
***subheader - cell0***
Cell ID: 0
Mutated Chromosome Sizes: {46: 194.55986862709122, 47: 149.786615867891, 48: 0.0006794090461994529, 49: 9.304206160720696e-05, 50: 1.0961534004739377, 51: 0.21780593898104428, 52: 2.038783714454979}
Intact Strands ID: 44, 0, 1, 2, 3, 5, 6, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45
Total DSB Count: 47
Total Misrepair Count: 43


In [3]:
header = {'SDD Version': 'SDDv1.0', 'Software': 'McMahon Empirical Model. v0.3', 'Author': 'Stephen J McMahonstephen.mcmahon@qub.ac.uk12/06/2018McMahon et alScientific Reports 10790', 'Simulation details': 'Empirical model of DNA damage distribution based on radial energy. Nucleus radius 4.32 um giving DSB energy of 60.1 keV/DSB on average', 'Source': 'Monoenergetic plane-parallel particle beam uniformly exposing nucleus. Energy: 0.975 MeV', 'Irradiation Target': '2.0', 'Additional Information': 'Generated for illustrative purposes for SDD Model manuscript. Data was generated in comprehensive data format.', 'Incident Particles': 2212, 'Mean Particle Energy': 0.975, 'Particle Fraction': 1.0, 'Dose or Fluence': [1.0, 1.0], 'DNA Density': 18.0630042979, 'Cell Cycle Phase': 0.0, 'DNA Structure': [0, 1], 'In vitro/in vivo': 0, 'Microenvironment': [20.0, 0.01], 'Damage definition': [0.0, 0.0, 10.0, -1.0, 0.0, 0.0], 'Damage and Primary Count': [3350, 3350], 'Data entries': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], 'Source Type': 1, 'Energy Distribution': [['M', '0']], 'World Volume': ['0', 5.0, 5.0, 5.0, 0.0, 0.0, 0.0], 'Scoring Volume': ['1', 4.32, 4.32, 4.32, 0.0, 0.0, 0.0], 'Proliferation status': [1, ''], 'Chromosomes': [46, [132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652, 132.608695652]], 'Time': 0.0}
header["Chromosomes"]

[46,
 [132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652,
  132.608695652]]

Author: Zhenlun Dai
Associated SDD File: n5MeV_inner_alpha_3_SDDOutput.txt
Old Chromosome Sizes: 46, 263.2, 258.1, 212.4, 203.4, 192.2, 181.8, 168.8, 155.5, 149.4, 144.0, 142.9, 140.8, 121.3, 113.0, 106.9, 94.7, 83.5, 81.0, 68.0, 66.6, 50.0, 52.6, 263.2, 258.1, 212.4, 203.4, 192.2, 181.8, 168.8, 155.5, 149.4, 144.0, 142.9, 140.8, 121.3, 113.0, 106.9, 94.7, 83.5, 81.0, 68.0, 66.6, 50.0, 52.6, 61.6, 165.2,
***end of master header***
***subheader - cell0***
Cell ID: 0
New Chromosome Sizes: {0: 263.2, 1: 258.1, 2: 212.4, 3: 203.4, 4: 181.8, 5: 168.8, 6: 149.4, 7: 144.0, 8: 142.9, 9: 140.8, 10: 121.3, 11: 113.0, 12: 106.9, 13: 94.7, 14: 83.5, 15: 81.0, 16: 68.0, 17: 66.6, 18: 50.0, 19: 52.6, 20: 263.2, 21: 258.1, 22: 212.4, 23: 203.4, 24: 192.2, 25: 181.8, 26: 168.8, 27: 155.5, 28: 149.4, 29: 144.0, 30: 142.9, 31: 140.8, 32: 121.3, 33: 113.0, 34: 106.9, 35: 94.7, 36: 83.5, 37: 81.0, 38: 68.0, 39: 66.6, 40: 50.0, 41: 52.6, 42: 61.6, 43: 165.2, 44: 192.2, 45: 136.81472756516587, 46: 0.00011975710900458125, 47: 0.001517230450232887, 48: 1.842417049147116e-06, 49: 2.0385755213270045, 50: 9.212085387844127e-07, 51: 0.00011699348341664972, 52: 4.14543838900272e-05, 53: 4.606042651289499e-06, 54: 9.304206160720696e-05, 55: 0.1316056931279661, 56: 16.296948119075836, 57: 0.21624725414692136}
Intact Strands ID: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43]
Total DSB Count: 47
Total Misrepair Count: 43
Medras-MC Log: ['0', '47', '0', '43', '10', '0', '0', '1', '0', '1', '0']
***data - cell0***
0; 0; 0/0/263.2/1; 1;
0; 1; 1/0/258.1/1; 1;
0; 2; 2/0/212.4/1; 1;
0; 3; 3/0/203.4/1; 1;
0; 4; 5/0/181.8/1; 1;
0; 5; 6/0/168.8/1; 1;
0; 6; 8/0/149.4/1; 1;
0; 7; 9/0/144.0/1; 1;
0; 8; 10/0/142.9/1; 1;
0; 9; 11/0/140.8/1; 1;
0; 10; 12/0/121.3/1; 1;
0; 11; 13/0/113.0/1; 1;
0; 12; 14/0/106.9/1; 1;
0; 13; 15/0/94.7/1; 1;
0; 14; 16/0/83.5/1; 1;
0; 15; 17/0/81.0/1; 1;
0; 16; 18/0/68.0/1; 1;
0; 17; 19/0/66.6/1; 1;
0; 18; 20/0/50.0/1; 1;
0; 19; 21/0/52.6/1; 1;
0; 20; 22/0/263.2/1; 1;
0; 21; 23/0/258.1/1; 1;
0; 22; 24/0/212.4/1; 1;
0; 23; 25/0/203.4/1; 1;
0; 24; 26/0/192.2/1; 1;
0; 25; 27/0/181.8/1; 1;
0; 26; 28/0/168.8/1; 1;
0; 27; 29/0/155.5/1; 1;
0; 28; 30/0/149.4/1; 1;
0; 29; 31/0/144.0/1; 1;
0; 30; 32/0/142.9/1; 1;
0; 31; 33/0/140.8/1; 1;
0; 32; 34/0/121.3/1; 1;
0; 33; 35/0/113.0/1; 1;
0; 34; 36/0/106.9/1; 1;
0; 35; 37/0/94.7/1; 1;
0; 36; 38/0/83.5/1; 1;
0; 37; 39/0/81.0/1; 1;
0; 38; 40/0/68.0/1; 1;
0; 39; 41/0/66.6/1; 1;
0; 40; 42/0/50.0/1; 1;
0; 41; 43/0/52.6/1; 1;
0; 42; 44/0/61.6/1; 1;
0; 43; 45/0/165.2/1; 1;
0; 44; 4/0/70.20679426647001/0, 4/70.2072582300885/70.20679426647001/0, 4/70.2072582300885/192.2/1; 1;
0; 45; 7/0/68.51992348933649/0, 7/69.61485997334124/68.51992348933649/0, 7/81.5827700799763/71.8713584271327/1, 7/81.5827700799763/89.36842569905214/0, 7/95.9553808264218/95.95538635367298/0, 7/98.4738138714455/98.47383966528436/0, 7/98.4738111078199/96.08718273696682/0, 7/108.18376540876778/108.18375896030805/0, 7/109.11248547393365/108.3816787707346/0, 7/111.65200682464454/111.65114180983413/0, 7/108.3816787707346/108.38143096563982/0, 7/109.11266787322275/109.11248547393365/0, 7/108.18376540876778/108.38143096563982/0, 7/109.11266787322275/111.47270555983413/0, 7/111.47331447867299/111.47325367890996/0, 7/111.47332000592417/111.65114180983413/0, 7/111.47279860189573/111.47281334123223/0, 7/111.47324723045023/111.47325367890996/0, 7/111.47324538803318/111.47323893957346/0, 7/111.47281426244076/111.47323893957346/0, 7/111.47331908471564/111.47332000592417/0, 7/111.65200682464454/155.5/0; 1;
0; 46; 7/96.08706297985782/96.08718273696682/0; 0;
0; 47; 7/71.65343549466824/71.65495272511848/0; 0;
0; 48; 7/111.47324538803318/111.47324723045023/0; 0;
0; 49; 7/69.61485997334124/71.65343549466824/0; 0;
0; 50; 7/111.47281334123223/111.47281426244076/0; 0;
0; 51; 7/71.65495272511848/71.6550697186019/0; 0;
0; 52; 7/71.87131697274882/71.8713584271327/0; 0;
0; 53; 7/111.47331447867299/111.47331908471564/0; 0;
0; 54; 7/111.47270555983413/111.47279860189573/0; 0;
0; 55; 7/95.95545728672985/96.08706297985782/0; 0;
0; 56; 7/95.95538635367298/95.95545728672985/0, 7/89.3685979650474/89.36845702014217/0, 7/89.3684422808057/89.36842569905214/0, 7/95.95531726303317/95.9553808264218/0, 7/95.95520856042654/95.95531726303317/0, 7/95.95520856042654/89.3685979650474/0, 7/89.3684422808057/89.36845702014217/0, 7/98.4738138714455/98.4738111078199/0, 7/98.47383966528436/108.18375896030805/0; 0;
0; 57; 7/71.6550697186019/71.87131697274882/0; 0;


In [4]:



#print('Fidelity analysis')
#fide = medrasrepair.repairSimulation(sddpath,'Fidelity', 24)

# print('\n\nMisrepair spectrum analysis')
# medrasrepair.repairSimulation(sddpath,'Spectrum')

In [5]:

'''
individual break set field:
    0 is file name
    1 is break set
    2 is break count
    3 is misrepair (in DSB? in fractional?)
    4 is misrepair standard deviation
    5 is inter-chromosome rate

    print('File\tBreak Set\tBreak Count\tMisrepair\tStdev\tInter-Chromosome Rate', end='')


summary field:
    0 is file name
    1 is break set (summary)
    2 is total break count
    3 is complexity (unit? how is it counted?)
    4 is average break count
    5 is average break count's standard deviation
    6 is average misrepairs
    7 is misrepair standard deviation
    8 is file average (of what? error rate? whatever that is)

	smry = (fileName+'\tSummary\t'+str(totalBreaks)+'\t'+str(complexity)+'\t'+str(averageBreaks)+
		   '\t'+str(breakStdev)+'\t'+str(averageMisrep)+'\t'+str(misrepStdev)+'\t'+
		   str(fileAverages[0]) )
'''

'''
individual break set field:
    0 is break set (index that starts at 0)
    1 is number of breaks
    2 is residual (number of DSB that remains)
    3 is number of misrepairs (include large misrepairs)
    4 is number of large misrepairs (larger than 3 MBP)
    5 is inter-chromosome misrepairs
    6 is the # of dicentrics
    7 is the # of rings
    8 is the # of excess linear fragments
    9 is the # of total aberrations
    10 is the viability

'''

'\nindividual break set field:\n    0 is break set (index that starts at 0)\n    1 is number of breaks\n    2 is residual (number of DSB that remains)\n    3 is number of misrepairs (include large misrepairs)\n    4 is number of large misrepairs (larger than 3 MBP)\n    5 is inter-chromosome misrepairs\n    6 is the # of dicentrics\n    7 is the # of rings\n    8 is the # of excess linear fragments\n    9 is the # of total aberrations\n    10 is the viability\n\n'

In [6]:
# def run_simulation(sddp, mode):
#     # Create an in-memory text stream
#     memory_file = io.StringIO()

#     if mode == 'Fidelity':
#       # Redirect stdout to the in-memory stream
#       with redirect_stdout(memory_file):
#           medrasrepair.repairSimulation(sddpath,'Fidelity', 24)

#       # Reset the stream position to the beginning
#       memory_file.seek(0)

#       # Read the data from the in-memory stream
#       repair_data = [line.split() for line in memory_file]# if line[0].isdigit()]

#       # Close the in-memory stream
#       memory_file.close()

#       return repair_data

#     elif mode == 'Spectrum':
#       # Redirect stdout to the in-memory stream
#       with redirect_stdout(memory_file):
#           medrasrepair.repairSimulation(sddp, 'Spectrum')
#           #repair_data = medrasrepair.repairSimulation(sddp, 'Spectrum')

#       # Reset the stream position to the beginning
#       memory_file.seek(0)

#       # Read the data from the in-memory stream if the line isnt empty
#       #repair_data = [line for line in memory_file]
#       repair_data = [line.split() for line in memory_file]
#       #repair_data = [line.split() for line in memory_file if line[0].isdigit()]

#       # Close the in-memory stream
#       memory_file.close()

#       return repair_data

#     elif mode == "postRepairDNA":
#           # Redirect stdout to the in-memory stream
#       with redirect_stdout(memory_file):
#           medrasrepair.repairSimulation(sddp, 'postRepairDNA')
#           #repair_data = medrasrepair.repairSimulation(sddp, 'Spectrum')

#       # Reset the stream position to the beginning
#       memory_file.seek(0)

#       # Read the data from the in-memory stream if the line isnt empty
#       #repair_data = [line for line in memory_file]
#       repair_data = [line.split() for line in memory_file]
#       #repair_data = [line.split() for line in memory_file if line[0].isdigit()]

#       # Close the in-memory stream
#       memory_file.close()

#       return repair_data
#     else:
#       print('Invalid mode')
#       return None

# def getpath(sddp):
#   for f in os.listdir(sddp):
#     if f.endswith(".txt"):
#       return f.split('.')[0]